### ISSUE TITLE
Update operation results in a multi-typed entity

### ISSUE THREAD

**Issue by MericFeyz** [ScorpioBroker/ScorpioBroker#484]
Release 4.1.10, Java-Kafka AAIO. Happens after upserting one entity with type X, then updating the same entity with type Y twice. 
As a result you have an entity with:
```
"type": [
    <X>,
    <Y>
  ],
```
Interesting matter is that if you upsert with X, update it once with type Y, the resulting entity has type Y. But after another update whether with Y or X, you get the entity with the type above. 

Might sound confusing, so to reproduce (in order):

```
curl -XPOST http://<scorpiohost>/ngsi-ld/v1/entityOperations/upsert?options=update -H "Content-type: application/ld+json" -d '[
        {
        "@context": "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld",
        "id": "urn:filter-test:12345",
        "type": "https://industry-fusion.com/types/v0.9/filter",
        "https://industry-fusion.com/types/v0.9/state": {
          "type": "Property",
          "value": "OF"
        },
        "https://industry-fusion.com/types/v0.9/strength": {
          "type": "Property",
          "value": "1.0"
        },
        "https://industry-fusion.com/types/v0.9/hasCartridge": {
          "type": "Relationship",
          "object": "urn:filterCartridge-test:32345"
        }
      }
    ]'
```
Then run this twice. Same id but different type:
```
curl -XPOST http://<scorpiohost>/ngsi-ld/v1/entities/urn:filter-test:12345/attrs -H "Content-type: application/ld+json" -d '{
  "@context":"https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context.jsonld",
  "id":"urn:filter-test:12345",
  "type":"https://industry-fusion.com/types/v0.9/filter_test",
  "https://industry-fusion.com/types/v0.9/state":{
    "type":"Property",
    "value":"OFF"
  },
  "https://industry-fusion.com/types/v0.9/strength":{
    "type":"Property",
    "value":"1.0"
  },
  "https://industry-fusion.com/types/v0.9/hasCartridge":{
    "type":"Relationship",
    "object":"urn:filterCartridge-test:22345"
  }
}'
```

This results in the following object after querying it:

```
{
  "id": "urn:filter-test:12345",
  "type": [
    "https://industry-fusion.com/types/v0.9/filter_test",
    "https://industry-fusion.com/types/v0.9/filter"
  ],
  "https://industry-fusion.com/types/v0.9/hasCartridge": {
    "type": "Relationship",
    "object": "urn:filterCartridge-test:22345"
  },
  "https://industry-fusion.com/types/v0.9/state": {
    "type": "Property",
    "value": "OFF"
  },
  "https://industry-fusion.com/types/v0.9/strength": {
    "type": "Property",
    "value": "1.0"
  }
}
```
If you then update the entity twice with type Z then it gets appended too, but nothing happens if you run it once again for some reason.

Is this intended? Documentations mention that type is immutable with update operation but they might be outdated. More crucial is if and how we can disable this behavior. Thanks for your help!

**Comment by ScorpioBroker** [comment]
Hi,

the documentation is unfortunately outdated there, we are currently working on updating it. There was some changes in the spec. So types became mutable but you can only add types (also in a federated case multiple types for the same entity id will be merged). This is goes rather deep through the application logic, so unfortunately no there is no on off switch for this behaviour, works as intended. 
To my understanding there is also currently no movement in ngsi-ld towards making certain types somehow explicitly immutable or something like this.

**Comment by ScorpioBroker** [comment]
So slight correction to me previous works as intended ... options=update should similar add a type and not overwrite it. 
@brijeshnec  could you please have a look thanks.

**Comment by martin-p-bauer** [comment]
NGSI-LD added the support for multi-types in v1.5.1 - before that, type was immutable. After a long discussion, it was decided that types can be added over time, but not removed. 

The reason is that type maps to @type, which is defined in JSON-LD and we cannot redefine it. Type directly has a value and is not reified, i.e. you cannot add timestamps like createdAt, modifiedAt, deletedAt and observedAt. Thus, there cannot be a temporal evolution of type in the temporal case. This would mean that, on replacing the type an entity currently has, type would also require a hard update in the temporal evolution, meaning an entity would lose its previous type, also for the past, and thus consumers retrieving it at different times would see an entity with an inconsistent type. 

From a philosophical perspective, we see it in that way, that logically an entity always had all its types, but not all of them were known to the system from the beginning. Thus with an update of any kind, a type can be added, but not removed (i.e. the behaviour you saw on update is actually a bug). If an entity changes in such a way that it no longer has a certain type, it logically becomes a different entity. 

It is possible to completely delete an entity that had one type and recreate it, as a new entity, with another type (however, to make sure this is done consistently in the distributed case, it is required that all instances in the different brokers are deleted).

**Comment by MericFeyz** [comment]
Thanks for the response. We look at it more from a user experience standpoint especially considering the possibility of user errors. In this case our greatest worry is that a small typo when updating an entity might brick the entire entity as we currently have no way of removing a type from an entity, only adding it. There's the option of deleting and recreating as @martin-p-bauer put it, or with upsert/replace as the NGSI-LD documentation mentions, but that requires manual fixing from an administrator.

The issue ultimately lies with the mismatch between our use-case and Scorpio's NGSI-LD version, which was recently bumped up from 1.3 to 1.7 with the 4.0 release. We were previously using the 3.0.11 release of Scorpio but due to a bug of the notifications, which were fixed in 4.0.3, we had to update to the 4.x and subsequently we had also updated the core context version. Is there anything in plans that would support being able to change NGSI-LD versions in the same release?
